### PR TITLE
Empty, id-only content Payload types are not supported for Topic based Subscriptions

### PR DESCRIPTION
**Pull request by volodymyr-korzh**
Added support for `id-only` and `empty` payload content types for notification 
triggered by R5, R4B, and R4 back-ported topic subscriptions.

### CODE REVIEW

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java [hunk 1]
```diff
@@ -423,13 +425,23 @@ private CanonicalSubscription canonicalizeR4B(IBaseResource theSubscription) {
...
-			retVal.getTopicSubscription().setTopic(getCriteria(theSubscription));
+			CanonicalTopicSubscription topicSubscription = retVal.getTopicSubscription();
+			topicSubscription.setTopic(getCriteria(theSubscription));
 
-			// WIP STR5 support other content types
-			retVal.getTopicSubscription()
-					.setContent(org.hl7.fhir.r5.model.Subscription.SubscriptionPayloadContent.FULLRESOURCE);
 			retVal.setEndpointUrl(channel.getEndpoint());
 			retVal.setChannelType(getChannelType(subscription));
+
```
Comment (epeartree) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java:433]: np.  i'd use the same technique here

**Review by epeartree — APPROVED** [review]

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java [hunk 2]
```diff
@@ -320,29 +318,33 @@ private CanonicalSubscription canonicalizeR4(IBaseResource theSubscription) {
 			}
 
 			if (channel.hasExtension(SubscriptionConstants.SUBSCRIPTION_TOPIC_CHANNEL_HEARTBEAT_PERIOD_URL)) {
-				org.hl7.fhir.r4.model.Extension timeoutExtension = channel.getExtensionByUrl(
+				org.hl7.fhir.r4.model.Extension channelHertbeatPeriotUrlExtension = channel.getExtensionByUrl(
```
Comment (epeartree) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java:321]: type Hert -> Heart

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java (hunk 1 above)
Comment (volodymyr-korzh) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java:436]: Fixed

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/topic/SubscriptionTopicPayloadBuilder.java [hunk 3]
```diff
@@ -87,7 +88,49 @@ public IBaseBundle buildPayload(
 		return retval;
 	}
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/topic/SubscriptionTopicPayloadBuilder.java:93]: Fixed

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/topic/SubscriptionTopicUtil.java [hunk 4]
```diff
@@ -72,4 +73,17 @@ public static IBaseResource extractResourceFromBundle(FhirContext myFhirContext,
 				.findFirst()
 				.orElse(null);
 	}
+
+	/**
+	 * Checks if {@link ActiveSubscription} has EMPTY {@link org.hl7.fhir.r5.model.Subscription.SubscriptionPayloadContent}
+	 * Used for R5/R4B/R4 Notification Status object building.
+	 */
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/topic/SubscriptionTopicUtil.java:80]: Fixed

File: hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizerTest.java [hunk 5]
```diff
@@ -131,37 +142,72 @@ public void testR5() {
 		assertEquals(456, canonical.getMaxCount());
 	}
 
-	@Test
-	void testR4Backport() {
+	@ParameterizedTest
+	@ValueSource(strings = {"full-resource", "id-only", "empty"})
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizerTest.java:146]: removed

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/BundleBuilder.java [hunk 6]
```diff
@@ -225,13 +221,26 @@ private IPrimitiveType<?> addAndPopulateTransactionBundleEntryRequest(
...
-		myEntryRequestMethodChild.getMutator().setValue(request, method);
+		addRequestMethod(request, theHttpVerb);
 		return url;
 	}
 
+	/**
+	 * Adds an entry containing an update (UPDATE) request without the body of the resource.
+	 *
+	 * @param theResource The resource to update.
+	 */
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/util/BundleBuilder.java:233]: Added

**Pull request merged**